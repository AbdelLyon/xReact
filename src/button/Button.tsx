import {
  Button as NextUIButton,
  ButtonProps as NextUIButtonProps,
} from "@nextui-org/react";

import { cn } from "@/utils";

export type ButtonVariant =
  | "solid"
  | "bordered"
  | "light"
  | "flat"
  | "faded"
  | "shadow"
  | "ghost";

export type ButtonColor =
  | "default"
  | "primary"
  | "secondary"
  | "success"
  | "warning"
  | "danger";

export type ButtonSize = "sm" | "md" | "lg";

export interface ButtonProps extends NextUIButtonProps {
  onPress?: (e: any) => void;
  LinkComponent?: React.ComponentType<any>;
  customStyles?: {
    base?: string;
    beforeContent?: string;
    afterContent?: string;
    content?: string;
  };
}

export const Button = ({
  variant = "solid",
  fullWidth = false,
  isLoading = false,
  isDisabled = false,
  startContent,
  endContent,
  className = "",
  LinkComponent,
  customStyles = {
    base: "",
    beforeContent: "",
    afterContent: "",
    content: "",
  },
  href,
  children,
  target,
  rel,
  ...props
}: ButtonProps) => {
  const baseStyles = cn(
    "transition-all font-normal dark:bg-opacity-90",
    fullWidth && "w-full",
    isLoading && "opacity-50 cursor-not-allowed",
    customStyles.base,
    className,
  );

  const Content = () => (
    <>
      {startContent && (
        <span className={cn("mr-2", customStyles.beforeContent)}>
          {startContent}
        </span>
      )}
      <span className={customStyles.content}>{children}</span>
      {endContent && (
        <span className={cn("ml-2", customStyles.afterContent)}>
          {endContent}
        </span>
      )}
    </>
  );

  if (href && LinkComponent) {
    return (
      <NextUIButton
        {...props}
        as={LinkComponent}
        className={baseStyles}
        href={href}
        rel={target === "_blank" ? "noopener noreferrer" : rel}
        target={target}
        variant={variant}
      >
        <Content />
      </NextUIButton>
    );
  }

  return (
    <NextUIButton {...props} className={baseStyles} isDisabled={isDisabled}>
      <Content />
    </NextUIButton>
  );
};
