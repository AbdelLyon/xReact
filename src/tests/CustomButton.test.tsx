import { render } from "@testing-library/react";
import { describe, it, expect } from "vitest";

import { CustomButton } from "@/button/Button";

describe("CustomButton", () => {
  it("should match the snapshot with default props", () => {
    const { asFragment } = render(<CustomButton>Click me</CustomButton>);

    expect(asFragment()).toMatchSnapshot();
  });

  it("should match the snapshot with icon", () => {
    const { asFragment } = render(
      <CustomButton icon={<i>icon</i>} iconPosition="left">
        Click me
      </CustomButton>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should match the snapshot with different variant and size", () => {
    const { asFragment } = render(
      <CustomButton
        icon={<i>icon</i>}
        iconPosition="right"
        size="lg"
        variant="outline"
      >
        Click me
      </CustomButton>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
