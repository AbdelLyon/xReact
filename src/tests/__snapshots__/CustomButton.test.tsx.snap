// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CustomButton > should match the snapshot with default props 1`] = `
<DocumentFragment>
  <button
    class="items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:opacity-80 h-9 px-4 py-2 flex gap-2"
  >
    <span
      class="flex-1"
    >
      Click me
    </span>
  </button>
</DocumentFragment>
`;

exports[`CustomButton > should match the snapshot with different variant and size 1`] = `
<DocumentFragment>
  <button
    class="items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-11 rounded-md px-8 flex gap-2"
  >
    <span
      class="flex-1"
    >
      Click me
    </span>
    <i>
      icon
    </i>
  </button>
</DocumentFragment>
`;

exports[`CustomButton > should match the snapshot with icon 1`] = `
<DocumentFragment>
  <button
    class="items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:opacity-80 h-9 px-4 py-2 gap-2 flex flex-row-reverse"
  >
    <span
      class="flex-1"
    >
      Click me
    </span>
    <i>
      icon
    </i>
  </button>
</DocumentFragment>
`;
